File: lua/resistor-color-trio/resistor-color-trio.lua
--- @enum Color
local Color = {
	["black"] = 0,
	["brown"] = 1,
	["red"] = 2,
	["orange"] = 3,
	["yellow"] = 4,
	["green"] = 5,
	["blue"] = 6,
	["violet"] = 7,
	["grey"] = 8,
	["white"] = 9,
}

return {
	--- @param c1 Color
	--- @param c2 Color
	--- @param c3 Color
	--- @return number, 'ohms' | 'kiloohms'
	decode = function(c1, c2, c3)
		local value = (Color[c1] * 10 + Color[c2]) * (10 ^ Color[c3])

		if value > 1000 then
			return value / 1000, "kiloohms"
		else
			return value, "ohms"
		end
	end,
}

File: python/resistor-color-trio/resistor_color_trio.py
import math

COLOR_VALUES = {
    "black": 0,
    "brown": 1,
    "red": 2,
    "orange": 3,
    "yellow": 4,
    "green": 5,
    "blue": 6,
    "violet": 7,
    "grey": 8,
    "white": 9,
}

UNITS = ["ohms", "kiloohms", "megaohms", "gigaohms"]


def label(colors: list[str]):
    value = COLOR_VALUES[colors[0]] * 10 + COLOR_VALUES[colors[1]]
    zeroes = COLOR_VALUES[colors[2]]
    unit_index, zeroes = divmod(zeroes, 3)

    if unit_index > len(UNITS) - 1:
        zeroes += unit_index - (len(UNITS) - 1)
        unit_index = len(UNITS) - 1

    value *= 10**zeroes

    while value > 0 and value % 1000 == 0 and unit_index < len(UNITS) - 1:
        unit_index += 1
        zeroes += 3
        value /= 1000

    return f"{math.floor(value)} {UNITS[unit_index]}"
